### PR TITLE
add csi cinder driver

### PR DESCRIPTION
**Pull request by edisonxiang**
Add csi cinder driver. 
Mainly including following interfaces:
* CreateVolume
* DeleteVolume
* ControllerPublishVolume
* ControllerUnpublishVolume
* NodePublishVolume
* NodeunPublishVolume

### CODE REVIEW

File: cinder/controllerserver.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	glog.V(4).Infof("ControllerUnpublishVolume %s on %s", volumeID, instanceID)
+
+	return &csi.ControllerUnpublishVolumeResponse{}, nil
+}
+
+func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req *csi.ValidateVolumeCapabilitiesRequest) (*csi.ValidateVolumeCapabilitiesResponse, error) {
+	for _, cap := range req.GetVolumeCapabilities() {
+		if cap.GetAccessMode().GetMode() != csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER {
```
Comment (chakri-nelluri) [inline cinder/controllerserver.go]: We do not need to implement this. I have fixed the default ValidateVolumeCapabilities, which loops over csiDriver caps.
https://github.com/kubernetes-csi/drivers/blob/master/csi-common/controllerserver-default.go#L47

File: cinder/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# CSI Cinder driver
+
+## Usage:
+
+### Start Cinder driver
+```
+$ export OS_USERNAME=admin
+$ export OS_PASSWORD=password
```
Comment (chakri-nelluri) [inline cinder/README.md]: Is there a way to hide the password? This will become a security hole.

File: cinder/controllerserver.go (hunk 1 above)
Comment (edisonxiang) [reply cinder/controllerserver.go]: Ok,  I will remove it.

File: cinder/cinderdriver.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+func GetCinderDriver() *cinderDriver {
+	runOnce.Do(func() {
+		driver = &cinderDriver{}
+	})
+	return driver
+}
+
+func NewIdentityServer(d *csicommon.CSIDriver) *identityServer {
```
[the remark this answers is not in the document]
Comment (edisonxiang) [reply cinder/cinderdriver.go]: No problem.

File: cinder/README.md (hunk 2 above)
Comment (edisonxiang) [reply cinder/README.md]: Hi Chakri, by [this link]( https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloudconf). It stores password in conf file. How do you think if we need to add these env into extra config file or keep them in env? Looking forward your suggestions.

**Comment by edisonxiang** [comment]
@chakri-nelluri Thanks your review. I will add unit test with driver later.

File: cinder/README.md (hunk 2 above)
Comment (chakri-nelluri) [reply cinder/README.md]: @edisonxiang how about storing it in a config file and pass the config file? In environments like K8S, config file can be made a secret and passed to the driver securely.

Comment (edisonxiang) [reply cinder/README.md]: Thanks @chakri-nelluri I will try.

**Comment by edisonxiang** [comment]
TODO:
* use config file
* add unit tests

**Comment by chakri-nelluri** [comment]
@edisonxiang Thanks for quick update. Please fix the TODO. There is a strong push back to stop accepting any new drivers with TODO's. We are also adding unit tests for existing drivers.

**Comment by edisonxiang** [comment]
Hi @chakri-nelluri @lpabon , unit tests have been added. PTAL, Thank you!

**Comment by rootfs** [comment]
please add license block in files that don't have yet.

**Comment by edisonxiang** [comment]
@rootfs I will add it at once. Thanks.

File: pkg/cinder/openstack/openstack.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+type OpenStack struct {
+	compute      *gophercloud.ServiceClient
+	blockstorage *gophercloud.ServiceClient
+}
+
+var OsInstance IOpenStack = nil
+
+func GetOpenStackProvider() (IOpenStack, error) {
```
Comment (rootfs) [inline pkg/cinder/openstack/openstack.go]: it can come in a followup pr but need to support reading config from config file (as in k8s)

Comment (edisonxiang) [reply pkg/cinder/openstack/openstack.go]: Yeah, I will

**Pull request merged**